File: daggercore/src/main/java/com/worldsnas/daggercore/navigator/changehandler/ArcFadeMoveChangeHandler.java
package com.worldsnas.daggercore.navigator.changehandler;

import android.annotation.TargetApi;
import android.os.Build;
import android.os.Bundle;
import android.transition.*;
import android.transition.Transition.TransitionListener;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.bluelinelabs.conductor.changehandler.SharedElementTransitionChangeHandler;
import com.bluelinelabs.conductor.internal.TransitionUtils;

import java.util.ArrayList;
import java.util.Collections;

@TargetApi(Build.VERSION_CODES.LOLLIPOP)
public class ArcFadeMoveChangeHandler extends SharedElementTransitionChangeHandler {

    private static final String KEY_SHARED_ELEMENT_NAMES = "ArcFadeMoveChangeHandler.sharedElementNames";

    private final ArrayList<String> sharedElementNames = new ArrayList<>();

    public ArcFadeMoveChangeHandler() { }

    public ArcFadeMoveChangeHandler(String... sharedElementNames) {
        Collections.addAll(this.sharedElementNames, sharedElementNames);
    }

    @Override
    public void saveToBundle(@NonNull Bundle bundle) {
        super.saveToBundle(bundle);

        bundle.putStringArrayList(KEY_SHARED_ELEMENT_NAMES, sharedElementNames);
    }

    @Override
    public void restoreFromBundle(@NonNull Bundle bundle) {
        super.restoreFromBundle(bundle);

        sharedElementNames.addAll(bundle.getStringArrayList(KEY_SHARED_ELEMENT_NAMES));
    }

    @Nullable
    @Override
    public Transition getExitTransition(@NonNull ViewGroup container, @Nullable View from, @Nullable View to, boolean isPush) {
        return new Fade(Fade.OUT);
    }

    @Nullable
    @Override
    public Transition getSharedElementTransition(@NonNull ViewGroup container, @Nullable final View from, @Nullable View to, boolean isPush) {
        Transition transition = new TransitionSet().addTransition(new ChangeBounds()).addTransition(new ChangeClipBounds()).addTransition(new ChangeTransform());
        transition.setPathMotion(new ArcMotion());

        // The framework doesn't totally fade out the "from" shared element, so we'll hide it manually once it's safe.
        transition.addListener(new TransitionListener() {
            @Override
            public void onTransitionStart(Transition transition) {
                if (from != null) {
                    for (String name : sharedElementNames) {
                        View namedView = TransitionUtils.findNamedView(from, name);
                        if (namedView != null) {
                            namedView.setVisibility(View.INVISIBLE);
                        }
                    }
                }
            }

            @Override
            public void onTransitionEnd(Transition transition) { }

            @Override
            public void onTransitionCancel(Transition transition) { }

            @Override
            public void onTransitionPause(Transition transition) { }

            @Override
            public void onTransitionResume(Transition transition) { }
        });

        return transition;
    }

    @Nullable
    @Override
    public Transition getEnterTransition(@NonNull ViewGroup container, @Nullable View from, @Nullable View to, boolean isPush) {
        return new Fade(Fade.IN);
    }

    @Override
    public void configureSharedElements(@NonNull ViewGroup container, @Nullable View from, @Nullable View to, boolean isPush) {
        for (String name : sharedElementNames) {
            addSharedElement(name);
        }
    }

    @Override
    public boolean allowTransitionOverlap(boolean isPush) {
        return false;
    }
}

File: navigation/src/androidMain/kotlin/com/worldsnas/navigation/changehandler/ArcFadeMoveChangeHandler.kt
package com.worldsnas.navigation.changehandler

import android.annotation.TargetApi
import android.os.Build
import android.os.Bundle
import android.transition.*
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Keep
import com.bluelinelabs.conductor.changehandler.SharedElementTransitionChangeHandler
import com.bluelinelabs.conductor.internal.TransitionUtils
import java.util.*

@TargetApi(Build.VERSION_CODES.LOLLIPOP)
class ArcFadeMoveChangeHandler : SharedElementTransitionChangeHandler {
    private val sharedElementNames = ArrayList<String>()

    @Keep
    constructor() {
    }

    constructor(vararg sharedElements: String) {
        Collections.addAll(this.sharedElementNames, *sharedElements)
    }

    override fun saveToBundle(bundle: Bundle) {
        super.saveToBundle(bundle)
        bundle.putStringArrayList(
            KEY_SHARED_ELEMENT_NAMES,
            sharedElementNames
        )
    }

    override fun restoreFromBundle(bundle: Bundle) {
        super.restoreFromBundle(bundle)
        sharedElementNames.addAll(bundle.getStringArrayList(KEY_SHARED_ELEMENT_NAMES))
    }

    override fun getExitTransition(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean
    ): Transition? {
        return Fade(Fade.OUT)
    }

    override fun getSharedElementTransition(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean
    ): Transition? {
        val transition: Transition =
            TransitionSet().addTransition(ChangeBounds())
                .addTransition(ChangeClipBounds())
                .addTransition(ChangeTransform())
        transition.pathMotion = ArcMotion()

        // The framework doesn't totally fade out the "from" shared element, so we'll hide it manually once it's safe.
        transition.addListener(object : Transition.TransitionListener {
            override fun onTransitionStart(transition: Transition) {
                if (from != null) {
                    for (name in sharedElementNames) {
                        val namedView =
                            TransitionUtils.findNamedView(
                                from,
                                name
                            )
                        if (namedView != null) {
                            namedView.visibility = View.INVISIBLE
                        }
                    }
                }
            }

            override fun onTransitionEnd(transition: Transition) {}
            override fun onTransitionCancel(transition: Transition) {}
            override fun onTransitionPause(transition: Transition) {}
            override fun onTransitionResume(transition: Transition) {}
        })
        return transition
    }

    override fun getEnterTransition(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean
    ): Transition? {
        return Fade(Fade.IN)
    }

    override fun configureSharedElements(
        container: ViewGroup,
        from: View?,
        to: View?,
        isPush: Boolean
    ) {
        for (name in sharedElementNames) {
            addSharedElement(name)
        }
    }

    override fun allowTransitionOverlap(isPush: Boolean): Boolean {
        return false
    }

    companion object {
        private const val KEY_SHARED_ELEMENT_NAMES =
            "ArcFadeMoveChangeHandler.sharedElementNames"
    }
}